sound/soc/sof/sof-client.c

```diff
@@ -0,0 +1,53 @@
+// SPDX-License-Identifier: (GPL-2.0 OR BSD-3-Clause)
+//
+// Copyright(c) 2020 Intel Corporation. All rights reserved.
+//
+// Author: Ranjani Sridharan <ranjani.sridharan@linux.intel.com>
+//
+
+#include <linux/device.h>
+#include "sof-client.h"
+
+static LIST_HEAD(client_list);
+static DEFINE_MUTEX(sof_client_mutex);
+
+int sof_client_drv_register(const struct sof_client_drv *drv, struct device *dev)
+{
+	struct sof_client_dev *cdev;
+
+	/* probe ops is mandatory for client drv */
+	if (!drv->ops->probe)
+		return -EINVAL;
+
+	cdev = devm_kzalloc(dev, sizeof(*cdev), GFP_KERNEL);
+	if (!cdev)
+		return -ENOMEM;
+
+	cdev->dev = dev;
+	cdev->drv = drv;
+
+	mutex_lock(&sof_client_mutex);
+	list_add(&cdev->list, &client_list);
+	mutex_unlock(&sof_client_mutex);
+
+	dev_dbg(dev, "%s SOF client driver registered\n", drv->name);
+
+	return 0;
+}
+EXPORT_SYMBOL_NS(sof_client_drv_register, SND_SOC_SOF_CLIENT);
+
+void sof_client_drv_unregister(const struct sof_client_drv *drv)
+{
+	struct sof_client_dev *cdev, *_cdev;
+
+	mutex_lock(&sof_client_mutex);
+	list_for_each_entry_safe(cdev, _cdev, &client_list, list) {
+		if (cdev->drv == drv) {
```

Comment:
Your `client_list` is global, and as far as I understand we can in theory at some point get multiple DSPs. Would they then possibly use the same driver? Then this matching would break?

Reply:
@lyakh I'm not sure about this case. But my guess would be that if there are multiple DSP, they possibly should not share the same driver. @lgirdwood could you please chime in?

Reply:
the same driver can be used for multiple devices, but each would need to have its own context and not a shared variable.

Reply:
@plbossart How would that work though? If you look at sof_client_drv_register() today, we create a client device for each client drv that is registered. If multiple devices share the same driver, then we probably need some kind of id table just like the virtbus driver, right?

Reply:
We can create a new global context (which can hold lists etc) per detected DSP device, i.e. on PCI/ACPI probe.